typing/typecore.ml

```diff
@@ -1723,20 +1721,23 @@ and type_pat_aux
       let initial_module_variables = !module_variables in
       let equation_level = !gadt_equations_level in
       let outter_lev = get_current_level () in
-      (* introduce a new scope *)
-      begin_def ();
-      let lev = get_current_level () in
-      gadt_equations_level := Some lev;
-      let type_pat_rec env sp = type_pat category sp expected_ty ~env in
-      let env1 = ref !env in
-      let p1 = type_pat_rec env1 sp1 in
-      let p1_variables = !pattern_variables in
-      let p1_module_variables = !module_variables in
-      pattern_variables := initial_pattern_variables;
-      module_variables := initial_module_variables;
-      let env2 = ref !env in
-      let p2 = type_pat_rec env2 sp2 in
-      end_def ();
+      (* Introduce a new scope using with_local_level without generalizations *)
+      let env1, p1, p1_variables, p1_module_variables, env2, p2 =
+        with_local_level begin fun () ->
+          let lev = get_current_level () in
+          gadt_equations_level := Some lev;
+          let type_pat_rec env sp = type_pat category sp expected_ty ~env in
+          let env1 = ref !env in
+          let p1 = type_pat_rec env1 sp1 in
+          let p1_variables = !pattern_variables in
+          let p1_module_variables = !module_variables in
+          pattern_variables := initial_pattern_variables;
+          module_variables := initial_module_variables;
+          let env2 = ref !env in
+          let p2 = type_pat_rec env2 sp2 in
+          (env1, p1, p1_variables, p1_module_variables, env2, p2)
+        end
+      in
       gadt_equations_level := equation_level;
       let p2_variables = !pattern_variables in
```

Comment:
I'm not fond of the structure of the code here -- the way that the equation level is handled, and the pattern variables collected, could be nicer -- but it is orthogonal to the current PR, we can try to remember to clean it up in a later PR.